snapd-glib/snapd-notices-monitor.c

```diff
@@ -0,0 +1,185 @@
+
+static void monitor_cb(SnapdClient* source,
+                       GAsyncResult* res,
+                       SnapdNoticesMonitor *self) {
+
+    g_autoptr(GError) error = NULL;
+    g_autoptr(GPtrArray) notices = snapd_client_get_notices_finish(source, res, &error);
+
+    if (g_cancellable_is_cancelled(self->cancellable)) {
+        g_clear_object(&self->cancellable);
+        g_clear_pointer(&self->last_date_time, g_date_time_unref);
+        g_clear_object(&self->last_notice);
+        return;
+    }
+
+    if (error != NULL) {
+        g_clear_object(&self->cancellable);
+        g_clear_pointer(&self->last_date_time, g_date_time_unref);
+        g_clear_object(&self->last_notice);
+        g_signal_emit_by_name(self, "error-event", error);
+        return;
+    }
+
+    if ((error == NULL) && (notices != NULL)) {
+        for (int i = 0; i < notices->len; i++) {
+            g_autoptr(SnapdNotice) notice = g_object_ref(notices->pdata[i]);
+            g_autoptr(GDateTime) last_occurred = (GDateTime *)snapd_notice_get_last_occurred(notice);
+
+            if (self->last_date_time == NULL || g_date_time_compare(self->last_date_time, last_occurred) <= 0) {
+                g_clear_pointer (&self->last_date_time, g_date_time_unref);
+                self->last_date_time = g_date_time_ref(last_occurred);
+                g_clear_object(&self->last_notice);
+                self->last_notice = g_object_ref(notice);
+            }
+            g_signal_emit_by_name(self, "notice-event", notice);
+        }
+    }
+    begin_monitor(self); // continue waiting for new notifications
+}
+
+static void begin_monitor(SnapdNoticesMonitor *self) {
+    snapd_client_notices_set_after_notice (self->client, self->last_notice);
+    snapd_client_get_notices_async(self->client,
+                                   self->last_date_time,
+                                   2000000000000000, // "infinity" (there is a limit in snapd, around 9 billion seconds)
+                                   self->cancellable,
+                                   (GAsyncReadyCallback)monitor_cb,
+                                   self);
+}
+
+/**
+ * snapd_notices_monitor_start:
+ * @monitor: a #SnapdNoticesMonitor
+ *
+ * Starts the asynchronous listening proccess, that will wait for new
+ * notices and emit a "notice-event" signal with the new notice as
+ * parameter.
+ *
+ * @return: FALSE if there was an error, TRUE if everything worked fine
+ * and the object is listening for events.
+ */
+gboolean snapd_notices_monitor_start(SnapdNoticesMonitor *self, GError **error) {
+
+    g_return_val_if_fail((error == NULL) || (*error == NULL), FALSE);
+    g_return_val_if_fail(SNAPD_IS_NOTICES_MONITOR(self), FALSE);
+    if (self->cancellable != NULL) {
+        *error = g_error_new(SNAPD_ERROR, SNAPD_ERROR_ALREADY_RUNNING, "The notices monitor is already running.");
+        return FALSE;
+    }
+    self->cancellable = g_cancellable_new();
```

Comment:
I would put this in init - then it's guaranteed to be available for the whole lifetime of the object.

Reply:
I think that it's better to keep it that way because the existence of the cancellable allows to know if the monitor is running or not.